ttnn/cpp/ttnn/operations/matmul/device/matmul_op.hpp

```diff
@@ -79,6 +81,25 @@ operation::ProgramWithCallbacks matmul_multi_core_reuse_mcast_2d_optimized(
     bool transpose_mcast,
     std::optional<UnaryWithParam> fused_activation,
     bool untilize_out);
+operation::ProgramWithCallbacks matmul_multi_core_reuse_mcast_2d_optimized_helper(
+    tt::tt_metal::Program& program,
```

Comment:
maybe create a few structures for input instead of haveing function which takes 20+ different args?

Reply:
Agreed - it would be good to get an explanation why all these args ar needed (and why and we just pass in the matmul op config which is used elsewhere in the PR anyways>

Reply:
I think this is just to not change matmul_multi_core_reuse_mcast_2d_optimized it's being duplicated.

Wouldn't you be able to add another parameter to matmul_multi_core_reuse_mcast_2d_optimized with default value of std::nullopt?

Also, this code will need to be converted at some point to use the new framework (tmp, https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/adding_new_ttnn_operation.html)
If any changes are made, it'd be good to keep in mind to try to not produce more work when that happens.

Reply:
@dmakoviichuk-tt I tried to follow the same structure as the existing file, and that's why the diff shows so many args.

@bbradelTT Here, the `program` and the `MatmulFusedOpSignaler` need to be passed in by reference for the fused op. But if they are passed by reference, I believe I cannot use `std::nullopt` (please correct me if I'm wrong).

As such, to avoid changing the existing call to `matmul_multi_core_reuse_mcast_2d_optimized`, I created a helper function `matmul_multi_core_reuse_mcast_2d_optimized_helper` that is a duplicate, but is sure to take in a reference to a `program` and `MatmulFusedOpSignaler`.

Then, `matmul_multi_core_reuse_mcast_2d_optimized` creates a new `program` and `MatmulFusedOpSignaler` inside as an empty optional object, which then get's passed along (see code [here](https://github.com/tenstorrent/tt-metal/pull/11760/files/ec96bfa1c1c03da3204b3d23836015c8ee81c88d#diff-962be7a139cf3d54d6bd165664db9a6f14173b4a6b74534ca63ce6ddbb37940dR1409)).

Another option is to use pointers instead of references. **Would this be preferred?** It would also remove the need to create the helper function.

Reply:
I was also having issue with implementing a pointer version, where `create_program_mcast_in0_in1` just creates a new `Program` object if an input `Program` pointer is a nullptr. However, doing something like this causes an error

```
// Take Program* program_ as input

Program program;
if (program_ != nullptr) {
    program = *program_; // compile error here
} else {
    program = Program();
}
```
I could also create a duplicate `Program`, and then just select which pointer to use (pointer to `Program` created in the function vs taken as input), but that would require changing all the subsequent uses of `Program` to be a pointer version (and creates a duplicate `Program` object).

Also as a clarification, this is why program and matmul_op_signaler need to be passed in by reference:
- The program is reused by all the ops being fused, ie, matmul and all_gather in this case (and datacopy, if enabled)
- the reference to the signaler struct is used to carry information from the matmul program factory to the all_gather program factory. Specifically, there are semaphores created in the matmul program factory and their ids need to be used by the all_gather kernels.

Reply:
So I see that you based your new function off the old one but I think ideally so many (effectively untyped) args should be avoided. Now I don't think it should be required that the already existing function be updated to use some consolidated args struct but I do still see value in moving to that for any new functions at the least (old functions can be updated over time).

Re the program question lifetime - I don't have a strong opinion on this. I think it's fine to have a "_detail" or "_impl" function (which would be ~ equivalent to the old function) just start taking a program by reference and making the caller decide the program lifetime (for fused wrapper, it's owned by fused op, for "old" matmul function, it's created there).